# 예제 5-1 누락 데이터 확인 

### 라이브러리 불러오기 

In [1]:
import seaborn as sns

### titanic 데이터셋 불러오기 

In [5]:
df = sns.load_dataset('titanic')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 63.0+ KB


RangeIndex을 보면 각 열에 891개의 데이터가 있다. deck열에는 203개의 non-null 데이터가 있으므로 누락 데이터가 688개라는 사실을 알 수 있다.

### deck 열의 NaN 개수 확인하기 

In [9]:
nan_deck = df['deck'].value_counts(dropna=False)
nan_deck

NaN    688
C       59
B       47
D       33
E       32
A       15
F       13
G        4
Name: deck, dtype: int64

dropna=False 옵션을 사용해여 NaN값을 포함한 데이터의 개수를 구할 수 있다.

### isnull() 메소드로 누락 데이터 찾기 

In [10]:
df.head().isnull()

survived  pclass    sex    age  sibsp  parch   fare  embarked  class  \
0     False   False  False  False  False  False  False     False  False   
1     False   False  False  False  False  False  False     False  False   
2     False   False  False  False  False  False  False     False  False   
3     False   False  False  False  False  False  False     False  False   
4     False   False  False  False  False  False  False     False  False   

     who  adult_male   deck  embark_town  alive  alone  
0  False       False   True        False  False  False  
1  False       False  False        False  False  False  
2  False       False   True        False  False  False  
3  False       False  False        False  False  False  
4  False       False   True        False  False  False

isnull() 메소드를 사용하면 NaN값에 True을 반환한다. deck열의 1행이 True이므로 NaN 값임을 알 수 있다.

### notnull() 메소드로 누락 데이터 찾기 

In [12]:
df.head().notnull()

survived  pclass   sex   age  sibsp  parch  fare  embarked  class   who  \
0      True    True  True  True   True   True  True      True   True  True   
1      True    True  True  True   True   True  True      True   True  True   
2      True    True  True  True   True   True  True      True   True  True   
3      True    True  True  True   True   True  True      True   True  True   
4      True    True  True  True   True   True  True      True   True  True   

   adult_male   deck  embark_town  alive  alone  
0        True  False         True   True   True  
1        True   True         True   True   True  
2        True  False         True   True   True  
3        True   True         True   True   True  
4        True  False         True   True   True

isnull() 메소드와 반대로 notnull()메소드는 NaN 값을 False로 반환한다.

### isnull() 메소드로 누락 데이터 개수 구하기 

In [16]:
df.isnull().sum(axis=0)

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

# 예제 5-2 누락 데이터 제거 

### for 반복문으로 각 열의 NaN 개수 계산하기 

In [17]:
missing_df = df.isnull()

In [19]:
for col in missing_df.columns:
    missing_count = missing_df[col].value_counts() # 각 열의 NaN 개수 파악
    try:
        print(col, ': ', missing_count[True]) # NaN값이 있으면 개수 출력
    except:
        print(col, ': ', 0) # NaN값이 없으면 0 출력

survived :  0
pclass :  0
sex :  0
age :  177
sibsp :  0
parch :  0
fare :  0
embarked :  2
class :  0
who :  0
adult_male :  0
deck :  688
embark_town :  2
alive :  0
alone :  0


### NaN값이 500개 이상인 열을 모두 삭제 - deck 열 

In [21]:
df_thresh = df.dropna(axis=1, thresh=500)
df_thresh.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'embark_town', 'alive',
       'alone'],
      dtype='object')

### age 열에 나이 데이터가 없는 모든 행 삭제 

In [22]:
df_age = df.dropna(subset=['age'], how='any', axis=0)
len(df_age)

714

subset을 age열로 한정하면 age열의 행 중에서 NaN값이 있는 모든 행(axis=0)을 삭제한다. 기본적으로 how='any' 옵션이 적용되어 있다. NaN값이 하나라도 존재하면 삭제한다. how='all' 옵션으로 입력하면 모든 데이터가 NaN값일 경우에만 삭제한다.

# 예제 5-3 평균으로 누락 데이터 바꾸기 

### age 열의 첫 10개 데이터 출력(5행에 NaN값) 

In [23]:
df['age'].head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: age, dtype: float64

### age열의 NaN값을 다른 나이 데이터의 평균으로 변경하기 

In [25]:
mean_age = df['age'].mean(axis=0) # age 열의 평균 계산(NaN값 제외)

In [26]:
df['age'].fillna(mean_age, inplace=True)

### age 열의 첫 10개 데이터 출력 

In [27]:
df['age'].head(10)

0    22.000000
1    38.000000
2    26.000000
3    35.000000
4    35.000000
5    29.699118
6    54.000000
7     2.000000
8    27.000000
9    14.000000
Name: age, dtype: float64

# 예제 5-4 가장 많이 나타나는 값으로 바꾸기 

### embark_town 열의 829행의 NaN 데이터 출력 

In [28]:
df['embark_town'][825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829            NaN
Name: embark_town, dtype: object

### embark_town 열의 NaN값을 승선도시 중에서 가장 많이 출현한 값으로 치환하기 

In [32]:
most_freq = df['embark_town'].value_counts(dropna=True).idxmax()
most_freq

'Southampton'

In [33]:
df['embark_town'].fillna(most_freq, inplace=True)

### embark_town 열 829행 데이터 확인

In [34]:
df['embark_town'][825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829    Southampton
Name: embark_town, dtype: object

# 예제 5-5 이웃하고 있는 값으로 바꾸기 

### embark_town 열의 NaN값을 바로 앞에 있는 828행의 값으로 변경하기 

In [ ]:
df['embark_town'].fillna(method='ffill')

### 

### 

### 